smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/endpointsV2/RuleSetParameterFinder.java

```diff
@@ -152,6 +153,64 @@ public Map<String, String> getContextParams(Shape operationInput) {
         return map;
     }
 
+    /**
+     * Get map of params to JavaScript equivalent of provided JMESPath expressions.
+     */
+    public Map<String, String> getOperationContextParamValues(OperationShape operation) {
+        Map<String, String> map = new HashMap<>();
+
+        Optional<OperationContextParamsTrait> trait = operation.getTrait(OperationContextParamsTrait.class);
+        if (trait.isPresent()) {
+            OperationContextParamsTrait operationContextParamsTrait = trait.get();
+            operationContextParamsTrait.getParameters().forEach((name, definition) -> {
+                String separator = ".";
+                String value = "this" + separator + "input";
+                String path = definition.getPath();
+
+                // Split JMESPath expression string on separator and add JavaScript equivalent.
+                for (String part : path.split("[" + separator + "]")) {
+                    if (value.endsWith(")")) {
+                        // The value is an object, which needs to run on map.
+                        value += ".map(obj => obj";
+                    }
+
+                    // Process keys https://jmespath.org/specification.html#keys
+                    if (part.startsWith("keys(")) {
+                        // Get provided object for which keys are to be extracted.
+                        String object = part.substring(5, part.length() - 1);
+                        value = "Object.keys(" + value + separator + object + ")";
+                        continue;
+                    }
+
+                    // Process list wildcard expression https://jmespath.org/specification.html#wildcard-expressions
+                    if (part.equals("*") || part.equals("[*]")) {
+                        value = "Object.values(" + value + ")";
+                        continue;
+                    }
+
+                    // Process hash wildcard expression https://jmespath.org/specification.html#wildcard-expressions
+                    if (part.endsWith("[*]")) {
+                        // Get key to run hash wildcard on.
+                        String key = part.substring(0, part.length() - 3);
+                        value = value + separator + key + separator + "map(obj => obj";
+                        continue;
+                    }
+
+                    // Treat remaining part as identifier without spaces https://jmespath.org/specification.html#identifiers
+                    value += separator + part;
+                }
+
+                // Close all open brackets.
+                value += ")".repeat((int) (
+                    value.chars().filter(ch -> ch == '(').count() - value.chars().filter(ch -> ch == ')').count()));
```

Comment:
would it be easier to increment the count as you write the opening parens?

Reply:
It's arguable. Incrementing the count as we write will require updating the count at multiple places.
It's easier to count at one place after the final string is ready.